onadata/apps/api/tools.py

```diff
@@ -760,3 +762,82 @@ def update_role_by_meta_xform_perms(xform):
+    decorator only caches GET responses and invalidates the cache entry when
+    any data modifying request is made(POST, PUT, PATCH, DELETE).
+
+    :param: cache_key_prefix : Prefix to append to the cache key used to store
+    the response
+
+    :param: cache_timeout : How long in seconds the cache should be valid for
+
+    :param: related_cache_entry_prefixes : Prefixes for cache entries that
+    might be affected or lose integrity when the decorated action executes
+    a modifying request.
+    """
+    def decorator(func):
+        def wrap(obj, request, *args, **kwargs):
+
+            def delete_cache_entries(
+                    request, lookup_value: Union[int, str] = None) -> str:
+                """
+                Deletes the cached response stored for the decorated function
+                and any other related cache entries if present.
+                """
+                main_cache_entry = get_cache_key(
+                    request, lookup_value=lookup_value)
+                cache.delete(main_cache_entry)
+
+                for related_key in related_cache_entry_prefixes:
+                    cache_key = get_cache_key(
+                        request, prefix=related_key, lookup_value=lookup_value)
+                    cache.delete(cache_key)
+
+            def get_cache_key(
+                    request,
+                    prefix: str = None,
+                    lookup_value: Union[int, str] = None) -> str:
+                """
+                Generates a unique key to be used for caching
+                """
+                cache_key = prefix if prefix else cache_key_prefix
+                if lookup_value:
+                    return f'{cache_key}{lookup_value}'
+                return cache_key + request.path
+
+            lookup_value = kwargs.get(
+                    getattr(obj, 'lookup_field', ""), None)
+            cache_key = get_cache_key(request, lookup_value=lookup_value)
+            if request.method == 'GET':
+                cached_response = cache.get(cache_key)
+
+                if cached_response:
+                    response = cached_response
+                else:
+                    response = func(obj, request, *args, **kwargs)
+
+                    if hasattr(response, 'render') and\
+                            callable(response.render):
+                        response.add_post_render_callback(
+                            lambda resp: cache.set(
+                                cache_key, resp, cache_timeout)
+                        )
+                    else:
+                        cache.set(cache_key, response, cache_timeout)
+            else:
+                # Invalidating cache in case the action modified any content
+                if request.method in ['POST', 'PUT', 'PATCH', 'DELETE']:
+                    delete_cache_entries(request, lookup_value)
+
+                obj.cache_response_key = cache_key
```

Comment:
Annotating the `self` argument with the `cache_key` to make it easier for users downstream to manually set their own cache entry for cases where the action has modified content.